File: 3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii/3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii.cs
public class Solution {
    public long CountOfSubstrings(string word, int k) {
         bool isvowel(char func)
        {
        
        if(func =='a')
                return true;
        if(func =='e')
                return true;
        if(func =='i')
                return true;
        if(func =='o')
                return true;
        if(func =='u')
                return true;
        if(func =='A')
                return true;
        if(func =='E')
                return true;
        if(func =='I')
                return true;
        if(func =='O')
                return true;
        if(func =='U')
                return true;
        else
                return false;
        }
        int[] indices  = new int[word.Length];
        int next1 = word.Length;
        for(int i = word.Length-1; i>=0; i--)
        {
            indices[i] = next1;
            if(!isvowel(word[i]))
                next1 = i;
        }
        int j = 0;
        int cons = 0; 
        Dictionary<char,int> mp = new Dictionary<char,int>();
        long ans = 0;
        for(int i = 0 ; i< word.Length; i++)
        {
            if(isvowel(word[i]))
                {
                    if(mp.ContainsKey(word[i]))
                        mp[word[i]]++;
                    else
                        mp.Add(word[i],1);
                }
            else
                cons++;
            if(cons>k)
            {
                while(j<=i && cons >k)
                {
                    if(isvowel(word[j]))
                    {
                        mp[word[j]]--;
                        if(mp[word[j]] == 0)
                            mp.Remove(word[j]);
                    }
                    else
                        cons--;
                    j++;
                }
                
            }
        while(j<i&&mp.Count() == 5 && cons == k)
            {
                int next = indices[i];
                ans += (next-i);
                if(isvowel(word[j]))
                {
                    if(--mp[word[j]] ==0)
                        mp.Remove(word[j]);
                }
                else
                    cons--;
                j++;
            }
            
        }
        return ans;
        
    }
}

File: 3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii/3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii.cpp
class Solution {
    bool isVowel(char ch) {
    ch = tolower(ch); // Convert to lowercase to handle uppercase letters
    return (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u');
}
public:
    long long countOfSubstrings(string word, int k) {
        long long ans = 0; 
        int cons = 0; 
        unordered_map<char,int>mp;
        int j = 0; 
        vector<int> consInd(word.size());
        consInd[consInd.size()-1]= consInd.size();
        int lastConsIdx = consInd.size();
        if(!isVowel(word[word.size()-1]))
             lastConsIdx = word.size()-1;
        
        for(int i = word.size()-2; i>=0;i--)
        {
            consInd[i]= lastConsIdx;
            if(!isVowel(word[i]))
                lastConsIdx = i; 
        }
        for(int i = 0; j< word.size(); )
        {
            if(isVowel(word[j]))
            {
                mp[word[j]]++;
            }
            else
                cons++;
            while(mp.size()==5 && cons>=k)
            {
                if(cons==k)
                {
                    int idx = consInd[j];
                    cout<<idx<<"  \n";
                    while(mp.size()==5 && cons == k)
                    {
                        ans += (idx-j);
                        if(isVowel(word[i]))
                        {
                            mp[word[i]]--;
                            if(mp[word[i]]==0)
                                mp.erase(word[i]);
                        }
                        else
                        {
                            cons--; 
                        }
                        i++;
                    }
                }
                else if(cons >k )
                {
                    while(cons > k)
                    {
                        if(isVowel(word[i]))
                        {
                            mp[word[i]]--;
                            if(mp[word[i]]==0)
                                mp.erase(word[i]);
                        }
                        else
                        {
                            cons--; 
                        }
                        i++;
                    }
                }
            }
            j++;
        }
        return ans; 
        
    }
};